### PR TITLE
Add license DSL as requested in #5540

### PR DESCRIPTION
**Pull request by MichelleNguy**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This PR adds a new DSL for license information as requested in #5540 and adds Licensee as a gem dependency.

### CODE REVIEW

**Comment by MichelleNguy** [comment]
@MikeMcQuaid 

I made an attempt at adding a `license` method and `licenses` method to the DSL. I was able to get it to work, except for when I validate the licenses using the `licensee` gem.

The `licensee` gem depends on rugged, which requires native extensions. It looks like homebrew depends on `portable-ruby` for it’s own ruby env and it needs all the gems to be vendored. I was able to vendor `licensee` but to do so I had to run `brew vendor-gems` with `HOMEBREW_NO_ENV_FILTERING=1` so that bundler could find `cmake `and compile the native extensions.

Do you have any guidance on how to vendor a gem that requires native extensions? Am I on the right track here?

**Comment by MikeMcQuaid** [comment]
Nice work so far @MichelleNguy!

> It looks like homebrew depends on `portable-ruby` for it’s own ruby env

Homebrew uses the system Ruby (or `portable-ruby` if the system Ruby is not 2.6.x).

 > and it needs all the gems to be vendored

Only gems used by `cmd`s (not `dev-cmd`s e.g. `dev-cmd/audit.rb` or `dev-cmd/style.rb`) need to be vendored. We cannot vendor gems that require native extensions.

> I was able to vendor `licensee` but to do so I had to run `brew vendor-gems` with `HOMEBREW_NO_ENV_FILTERING=1` so that bundler could find `cmake `and compile the native extensions.

I would suggest adding it to the `Gemfile` and not vendoring and as long as it can be installed by e.g. `brew style` or `brew audit` locally/on CI that's sufficient.

**Comment by MichelleNguy** [comment]
> I would suggest adding it to the `Gemfile` and not vendoring and as long as it can be installed by e.g. `brew style` or `brew audit` locally/on CI that's sufficient.

Thanks @MikeMcQuaid!

Can you provide more insight to `brew audit`? Would this install the gem? I've added `licensee` to the gemfile and ran `brew audit` 4+ hours later, it's still going.

**Comment by igas** [comment]
@MichelleNguy `brew audit` on its own will be checking all the formulas and it may take time. You can pass formula name to speed up the process. `brew audit curl` for example.

**Comment by MikeMcQuaid** [comment]
> Can you provide more insight to `brew audit`? Would this install the gem?

Yup, `brew audit` automatically installs all gems.

As for speed, as @igas mentioned I'd recommend testing with a single formula. If it's still taking a long time (>10m) then paste the output here and we'll try to help you debug. 

Thanks!

**Comment by MichelleNguy** [comment]
@MikeMcQuaid I'm struggling to get Homebrew to run with my system ruby (managed by rbenv). I've tried manually overriding `HOMEBREW_RUBY_PATH` and it seems brew is still using portable Ruby. 

I looked at `ruby.sh` and it seems that the default is to use portable Ruby if it is available. 

https://github.com/Homebrew/brew/blob/f527f03fb5d57ac19ac03b8968536bc3a3346849/Library/Homebrew/utils/ruby.sh#L14-L32

Do you know how I can force my system to not use portable Ruby?

**Comment by MikeMcQuaid** [comment]
> I'm struggling to get Homebrew to run with my system ruby (managed by rbenv).

@MichelleNguy `rbenv` does not manage the system Ruby. By system Ruby I mean the one at `/usr/bin/ruby`. It'll use that unless it is too old in which case it will use Portable Ruby (and you should let it). Can you post your `brew config` output so we can check?

**Comment by MichelleNguy** [comment]
>Can you post your `brew config` output so we can check?

@MikeMcQuaid 

Sure!

```HOMEBREW_VERSION: 2.2.10-79-gc079bdf
ORIGIN: https://github.com/Homebrew/brew
HEAD: c079bdf159c528a4c86a2eae006efc5914943f11
Last commit: 2 minutes ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: a30239f85eaed019c60bfdfd9b78ce32deaa9126
Core tap last commit: 2 days ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_DEV_CMD_RUN: 1
HOMEBREW_VISUAL: nano
CPU: quad-core 64-bit kabylake
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3/bin/ruby
Clang: 10.0 build 1001
Git: 2.25.0 => /usr/local/bin/git
Curl: 7.54.0 => /usr/bin/curl
macOS: 10.14.6-x86_64
CLT: 10.3.0.0.1.1562985497
Xcode: N/A
XQuartz: 2.7.11 => /opt/X11```

**Comment by MichelleNguy** [comment]
@MikeMcQuaid 

I think this PR should work, but when tests are run in CI, calling `brew tap` before the ruby gems are installed is causing an error: 

https://github.com/Homebrew/brew/blob/master/.github/workflows/tests.yml#L50-L52

produces:
`Error cannot load such file -- licensee`

https://github.com/Homebrew/brew/pull/7185/checks?check_run_id=517005127#step:4:39

From what I understood, you expect that a dev command would install the dependency, but it seems that calling `brew tap` in the CI script before the `licensee` dependency is installed causes the build to fail.

1. Does this require a change to CI?
2. Is there a way to make it so that my change doesn’t impact the `tap` command?

It looks like the `tap` command causes all of `brew` to be loaded, including my changes in the `DSL` which require the `licensee` gem.

File: Library/Homebrew/cask/dsl.rb [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 require "locale"
 require "lazy_object"
+require "licensee"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/dsl.rb]: You should avoid requiring/using `licensee` outside of `Library/Homebrew/dev-cmd/audit.rb` or `Library/Homebrew/cask/cmd/audit.rb`.

File: Library/Homebrew/cask/dsl.rb [hunk 2]
```diff
@@ -92,6 +95,30 @@ def name(*args)
       @name.concat(args.flatten)
     end
 
+    def license(*args)
+      @licenses ||= []
+
+      return @licenses unless args
+
+      valid_licenses = Licensee.licenses.map(&:key)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/dsl.rb]: I think it's worth just allowing any value of license to be specified and relying on `brew audit` and `brew cask audit` to flag invalid ones.

**Comment by MikeMcQuaid** [comment]
> Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3/bin/ruby

Yup, this means you're using the portable Ruby. Shouldn't be any problems here but you may be hitting an issue such as https://github.com/Homebrew/brew/issues/5068. I can't suggest much beyond upgrading to a newer macOS version 😭

**Comment by MikeMcQuaid** [comment]
Nice work so far @MichelleNguy 👏. To be clear: this stuff isn't easy and you're doing really great so far.

File: Library/Homebrew/cask/dsl.rb [hunk 3]
```diff
@@ -92,6 +94,25 @@ def name(*args)
       @name.concat(args.flatten)
     end
 
+    def license(*args)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/dsl.rb:97]: Would be good to have something similar in `formula.rb`; I think that's probably a priority over `cask/dsl.rb` (where many/most packages will have a proprietary license). Nice work so far!

**Comment by igas** [comment]
Hi @MichelleNguy, just want to check if you are still interested in it. Do you need any help with it?

**Comment by MichelleNguy** [comment]
@igas, thanks for the offer! I am indeed still working on this, though for now I think I am good. Should have some updates soon.

**Comment by MichelleNguy** [comment]
Hi @MikeMcQuaid and @igas,

You had mentioned in https://github.com/Homebrew/brew/pull/7185#issuecomment-600192901 that we cannot vendor native gems like `licensee` (rugged) and non-dev commands require vendored gems.

So the `cask audit` command is not a dev command and in our current approach requires `licensee`.

The regular `audit` command is a dev command and so we can use `licensee` with that.

Please advise for how to handle this.

**Comment by MikeMcQuaid** [comment]
> So the `cask audit` command is not a dev command and in our current approach requires `licensee`.

Either of these two options is fine:
1. treat `cask audit` as a dev-cmd
2. omit `cask audit` from the initial PR (as there's less need for the DSL in Casks)

**Comment by MichelleNguy** [comment]
My friend Shane is helping out with this PR. This would be his first dive into open source!

**Comment by zwick-stripe** [comment]
@MichelleNguy Thank you so much for all of your work on this! I'd like to see it finished and merged into the mainline branch, so please let me know how/if I can help get it across the finish line.

**Comment by MikeMcQuaid** [comment]
@zwick-stripe the current work on this is in https://github.com/Homebrew/brew/pull/7762

**Pull request closed without merging**